src/Database/LSMTree/Internal/Index.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE TypeFamilies #-}
+
+{-|
+    Provides a common interface to different types of fence pointer indexes and
+    their accumulators.
+-}
+module Database.LSMTree.Internal.Index
+(
+    Index (search, sizeInPages, fromSBS),
+    IndexAcc (ResultingIndex, appendSingle, appendMulti, unsafeEnd)
+)
+where
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.ByteString.Short (ShortByteString)
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.Chunk (Chunk)
+import           Database.LSMTree.Internal.Entry (NumEntries)
+import           Database.LSMTree.Internal.Page (NumPages, PageSpan)
+import           Database.LSMTree.Internal.Serialise (SerialisedKey)
+
+-- | The class of index types.
+class Index i where
+
+    {-|
+        Searches for a page span that contains a key–value pair with the given
+        key. If there is indeed such a pair, the result is the corresponding
+        page span; if there is no such pair, the result is an arbitrary but
+        valid page span.
+    -}
+    search :: SerialisedKey -> i -> PageSpan
+
+    -- | Yields the number of pages covered by an index.
+    sizeInPages :: i -> NumPages
+
+    {-|
+        Reads an index along with the number of entries of the respective run
+        from a byte string.
+
+        The byte string must contain the serialised index exactly, with no
+        leading or trailing space.
+    -}
+    fromSBS :: ShortByteString -> Either String (NumEntries, i)
+
+{-|
+    The class of index accumulator types, where an index accumulator denotes an
+    index under incremental construction.
+
+    Incremental index construction is only guaranteed to work correctly when the
+    following conditions are met:
+
+      * The provided information about pages is not in conflict with the
+        requirement that the keys in a run are ascending.
```

Comment:
> I deliberately avoided saying that keys should be appended in ascending order, because we actually don’t append keys but key ranges, where a key range is typically specified by a minimum and a maximum key.

Sorted ranges then?

> Yes, in our concrete index implementations, we store only the minimum or only the maximum keys, but the interface does not mandate this. Incidentally, I think it would be better for the interface to do so, but I fear it’s too late now to change this.

I'm not sure I understand what you mean here. For `search` to work properly, the stored key ranges have to be sorted. The implementation does not check this, but we've been mentioning it explicitly as a requirement for correctness

Reply:
> I’ll try to change the description to something direct.

Done in e7ef851269b3e37e4d89d8a21217ddb51f05fa26.

Reply:
> Done in https://github.com/IntersectMBO/lsm-tree/commit/e7ef851269b3e37e4d89d8a21217ddb51f05fa26.

LGTM

> What I meant is that it would not be necessary for appendSingle to receive a minimum and a maximum key of a page. Instead, it would be enough for it to only receive either a minimum or a maximum key and have indexes of all types store only minimum keys or only maximum keys. Actually, this would also better resemble the idea of a fence pointer index: the index should store fences, which are situated between pages, and a close approximation of a fence is either the maximum key of the preceding page or the minimum key of the succeeding page.

I understand what you mean, though it's not so straightforward for the compact index. The compact index needs to know about the both the minimum and maximum key on each page so that it can decide whether the 64-byte prefix of the minimum key is a proper fence. If the maximum key on the previous page and the minimum key on the next page have the same 64-byte prefix, then the minimum key is not a proper fence. This is also documented in the `IndexCompact`:

https://github.com/IntersectMBO/lsm-tree/blob/11738bdd7b352d32e43cd8e45d93fde391c3b91d/src/Database/LSMTree/Internal/IndexCompact.hs#L152-L179